### PR TITLE
[Codegen][GPU] Add pass to resolve scf.forall ops

### PR DESCRIPTION
**Pull request by qedawkins**
This pass differs from the upstream pattern in that it supports dynamic trip count forall loops as well as overprovisioning, however it has a few restrictions with respect to indexing. It only supports linearized thread indices, as well as loops where the iteration order matches the order of fastest changing thread index (i.e. the outer most loop will be assigned the slowest changing thread indices). These are practical choices to avoid overcomplicating the implementation of this pass by requiring us to normalize `scf.forall` operations before this pass. Such a normalization pass is TODO.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+  if (!llvm::all_of(mapping, [](Attribute attr) {
+        return cast<DeviceMappingAttrInterface>(attr).isLinearMapping();
+      })) {
+    forallOp.emitOpError("unimplemented: resolution of scf.forall ops without "
+                         "linear id mappings.");
+    return failure();
+  }
+
+  if (!forallOp.isNormalized()) {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp:75]: Its ok for now, but you could also normalize on distribution.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+  bool perfectlyDivides =
+      staticProducerCount &&
+      staticProducerCount.value() % flatTotalNumWorkers == 0;
+
+  // Step 3. Create the `scf.for` loop for the loop.
+  // If the workgroup count perfectly divides the loop's worker count, then we
+  // can use a lower bound of 0 and keep the loop bounds static. This helps
+  // simplify later loop folding patterns without an `affine.linearize_index` op
+  // to help with inferring int ranges.
+  Value lb = perfectlyDivides ? rewriter.create<arith::ConstantIndexOp>(loc, 0)
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp:117]: I think the mixing of the perfectly divides and not case makes the code hard to follow. Could you do a normalization after the fact that just does this all the time?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Mostly looks good!

Ill trade you review on this with review on https://github.com/iree-org/iree/pull/18368 . They are doing similar things... can we refactor the logic here (or I can do it on the other PR) to have utilities we can use and maybe eventually upstream?

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp (hunk 1 above)
Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp:75]: I was thinking we could just add the loop normalization pass immediately before this rather than complicating the code here. Either way is ok with me though.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp (hunk 2 above)
Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp:117]: I commented about this in an earlier PR that I don't have immediately on hand, but the short answer is that I don't see a short term way to avoid the extra logic here and loop normalization doesn't solve the problem. The code here generates 2 possible paths based on `perfectlyDivides`
```
// if perfectly divides is true
scf.for %c0 to %wg_size step %loop_trip_count

// else
scf.for %thread_id to %wg_size step %loop_trip_count
```

if `wg_size == %loop_trip_count`, then this will be
```
// if perfectly divides is true
scf.for %c0 to %wg_size step %wg_size

// else
scf.for %thread_id to %wg_size step %wg_size
```
The first loop can be trivially folded away, while the second one requires proving that `0 <= thread_id < %wg_size`, and then writing a folder for `scf.for` that can query for such range information, which requires a non-trivial amount of work to do properly. Instead just adding a few extra lines of logic in the resolution pattern (here) seems simpler to me because that's the point when we can most easily directly compare the workgroup size against the loop trip count.

Also as a side benefit, loops with constant iteration bounds is easier to analyze for loop unrolling.

**Comment by qedawkins** [comment]
> Mostly looks good!
> 
> Ill trade you review on this with review on #18368 . They are doing similar things... can we refactor the logic here (or I can do it on the other PR) to have utilities we can use and maybe eventually upstream?

Yes, refactoring and unifying sounds great to me. It'll be easier to refactor the other PR based on whichever one lands first, so let me go review your PR now. I started on the review but forgot to come back to it, sorry about that...

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp (hunk 1 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp:75]: Ok, makes sense.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp:117]: ok... still has the smell of a premature optimization but I see why you need it.

@krzysz00  maybe something for range inference to handle.

**Review by MaheshRavishankar — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp (hunk 2 above)
Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp:117]: +1 this is exactly the kind of thing that would be great for range inference to handle.

Comment (krzysz00) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp:117]: Might be good to have a chat about this at some point - especially since the inequalities here feel more like the value bounds stuff someone else did than the constant range stuff ... but perhaps if https://github.com/iree-org/iree/pull/17707 got pulled in some of this might rewrite away eventually?

Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp:117]: I think there are two parts here. 1 is #17707, which might be good enough on its own, but the other would be `affine.linearize_index` if the affine maps generated here are too unwieldy.
